### PR TITLE
refactor(op): polish generate_proposals_op

### PR DESCRIPTION
**Pull request by reyoung**
Polish styles in generate_proposals_op.

1. inline lambda functions rathar than use std::function to save var.
2. add `static inline` to template functions .cc
   * Make them static to prevent generating symbols.
   * Make them inline to give compiler a hit inline them as possible.
   * Not if the function is not static, they cannot be inlined since the
     symbols should be exported.
3. add `static` to global functions in .cc
   * Make them static to prevent generating symbols.
4. Use Vector<uint64> instead manually manange storage between devices.
5. Prefer to use platform::ForRange, so we can optimize `ForRange` by
   just changing `for_range.h` if it is needed.
6. Do not change shape of inputs

test=develop

### CODE REVIEW

File: paddle/fluid/operators/detection/generate_proposals_op.cc [hunk 1]
```diff
@@ -25,21 +27,17 @@ namespace operators {
...
-      : out_(out), offset_(offset), to_add_(to_add) {}
-
-  template <typename T>
-  void apply() const {
-    auto *out_data = out_->data<T>();
-    auto *to_add_data = to_add_->data<T>();
-    memcpy(out_data + offset_, to_add_data, to_add_->numel() * sizeof(T));
-  }
-};
+static void AppendProposals(Tensor *dst, int64_t offset, const Tensor &src) {
```
Comment (reyoung) [inline paddle/fluid/operators/detection/generate_proposals_op.cc:32]: It is no need to visit type if only SizeOfType is needed.

File: paddle/fluid/operators/detection/generate_proposals_op.cc [hunk 2]
```diff
@@ -185,24 +186,24 @@ void FilterBoxes(const platform::DeviceContext &ctx, Tensor *boxes,
...
-  return pair1.first > pair2.first;
-}
-
 template <class T>
-void GetMaxScoreIndex(const std::vector<T> &scores,
-                      std::vector<std::pair<T, int>> *sorted_indices) {
+static inline std::vector<std::pair<T, int>> GetSortedScoreIndex(
+    const std::vector<T> &scores) {
+  std::vector<std::pair<T, int>> sorted_indices;
+  sorted_indices.reserve(scores.size());
```
Comment (reyoung) [inline paddle/fluid/operators/detection/generate_proposals_op.cc:193]: reserve is better.

File: paddle/fluid/operators/detection/generate_proposals_op.cc [hunk 3]
```diff
@@ -392,10 +394,9 @@ class GenerateProposalsKernel : public framework::OpKernel<T> {
     for (int i = 0; i < scores_slice.numel(); ++i) {
       index[i] = i;
     }
-    std::function<bool(const int64_t &, const int64_t &)> compare =
```
Comment (reyoung) [inline paddle/fluid/operators/detection/generate_proposals_op.cc:395]: lambda的类型不是std::function，这里使用std::function接lambda，会导致

1. 有一次转型操作。
2.compare不能被inline

File: paddle/fluid/operators/detection/generate_proposals_op.cc [hunk 4]
```diff
@@ -352,32 +361,25 @@ class GenerateProposalsKernel : public framework::OpKernel<T> {
       scores_slice.Resize({h_score * w_score * c_score, 1});
 
       std::pair<Tensor, Tensor> tensor_pair =
-          ProposalForOneImage(dev_ctx, im_info_slice, *anchor, *var,
+          ProposalForOneImage(dev_ctx, im_info_slice, anchors, variances,
                               bbox_deltas_slice, scores_slice, pre_nms_top_n,
                               post_nms_top_n, nms_thresh, min_size, eta);
-      Tensor proposals = tensor_pair.first;
```
Comment (reyoung) [inline paddle/fluid/operators/detection/generate_proposals_op.cc:358]: 这里之前调用过复制构造函数。。

**Review by qingqing01 — APPROVED** [review]
Thanks very much.

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**